src/libstd/io/stdio.rs

```diff
@@ -175,13 +175,17 @@ impl Read for Stdin {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<'a> Read for StdinLock<'a> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        self.inner.read(buf)
+        self.inner.borrow_mut().read(buf)
     }
 }
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<'a> BufRead for StdinLock<'a> {
-    fn fill_buf(&mut self) -> io::Result<&[u8]> { self.inner.fill_buf() }
-    fn consume(&mut self, n: usize) { self.inner.consume(n) }
+    fn fill_buf(&mut self) -> io::Result<&[u8]> {
+        self.inner.borrow_mut().fill_buf()
```

Comment:
Hm no, nevermind, that's not valid. I... actually think that using a reentrant mutex may make this fundamentally unsafe. For example this means that you can acquire a `StdinLock` at any time:

```rust
let s = io::stdin();
let mut h1 = s.lock();
let mut h2 = s.lock();
let buf1 = h1.fill_buf();
let buf2 = h2.fill_buf();
// buf1/buf2 alias!
```